### PR TITLE
Set channel reopen option and retry options for robustness

### PR DESCRIPTION
**Pull request by jean-philippe-martin**
(these are all cloud-specific)

### CODE REVIEW

**Comment by jean-philippe-martin** [comment]
The failure is: "ERROR: (gcloud.components.update) The component [bq-nix] failed to download." This looks like a transient issue related to gcloud components update.

FWIW I was able to update my desktop's installation with no issue.

**Comment by jean-philippe-martin** [comment]
I knew it, this just needed a retry.

**Comment by jean-philippe-martin** [comment]
It'd be nice if we could please have a review of this code. The retries fix a real problem, and there isn't much code so it shouldn't be too hard to review.

**Comment by droazen** [comment]
@jean-philippe-martin I'm concerned that this fix does not set retries for all pathways into the NIO code. In particular, what about classes like `ReadsDataSource` that add a prefetcher via `SeekableByteChannelPrefetcher.addPrefetcher()`, or classes in htsjdk that create a `Path` from a raw `String`? Could you comment on this issue before I review? Thanks!

**Comment by jean-philippe-martin** [comment]
@droazen sure. This PR updates both `BucketUtils.getPathOnGcs` and `IOUtils.getPath` (the latter indirectly) to create a Google Cloud Storage filesystem with a default reopen set to 3. Anyone opening the given `Path` (or even a `Path` derived from it, e.g. via `subpath`) will have the retries enabled.

`addPrefetcher` wraps an existing `Path`, so it inherits the underlying `Path`'s retry behavior.

Classes within htsjdk that create a `Path` from a String, without using our utility code, would indeed not set retries and errors on those files would not get the benefit of our extra retry. The exception to that is the Spark code path with `NioBam`: this goes via `ReadsIterable`, which also sets the retries.

My understanding is that the normal way to open BAMs for tools always creates the `Path` object when parsing the command line, as in e.g. `ReadInputArgumentCollection::GetReadIndexPaths()`. That code path that calls `IOUtils.getPath` (and thus sets the retries). I would expect that the added support for VCF follows the same style, though it looks like it doesn't.

If there's a code path that I missed where you think we need retries then please let me know so I can add it!

**Comment by jean-philippe-martin** [comment]
Since the code isn't reviewed yet I took the liberty of adding one much push with a single change: adding a "synchronized" to protect against a potential data race in `SeekableByteChannelPrefetcher`. The contract for `ReadableByteChannel` (which this implements) requires `read` to be thread-safe.

I don't know whether this was the cause of #2516. I haven't been able to reproduce it since, but then again even before this change it wasn't easy to trigger.

**Comment by jean-philippe-martin** [comment]
Edit: that wasn't it, I saw #2516 again last night.

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java [hunk 1]
```diff
@@ -393,9 +391,28 @@ public static String getPathWithoutBucket(String path) {
         final String[] split = gcsUrl.split("/");
         final String BUCKET = split[2];
         final String pathWithoutBucket = String.join("/", Arrays.copyOfRange(split, 3, split.length));
-        return CloudStorageFileSystem.forBucket(BUCKET).getPath(pathWithoutBucket);
+        CloudStorageConfiguration cloudConfig = CloudStorageConfiguration.builder()
+            // if the channel errors out, re-open up to this many times
+            .maxChannelReopens(3)
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: Can you extract the number of retries as a `public static final` constant `BucketUtils.GCS_NIO_RETRIES`, and use that constant both here and in `ReadsIterable` below?

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java [hunk 2]
```diff
@@ -393,9 +391,28 @@ public static String getPathWithoutBucket(String path) {
...
+                .build())
+            .setRetryParams(RetryParams.newBuilder()
+                .setRetryMaxAttempts(10)
+                .setRetryMinAttempts(6)
+                .setMaxRetryDelayMillis(30000)
+                .setTotalRetryPeriodMillis(120000)
+                .setInitialRetryDelayMillis(250)
+                .build());
+    }
     // TODO(jpmartin): uncomment once shaded.cloud_nio.com.google.auth.Credentials is visible.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: The package name fix is still not released, correct?

**Comment by droazen** [comment]
Before merging this, could you have a look at `FeatureDataSource` (in particular, `getFeatureReader()`) to make sure it will also benefit from the new retry settings? This is the code path hit when reading VCFs.

**Comment by droazen** [comment]
Review complete, back to @jean-philippe-martin , squash and merge after addressing comments.

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java (hunk 2 above)
Comment (jean-philippe-martin) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: Actually there was a new release just two days ago. Updating to the newest gcloud should probably be a separate PR, though.

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java (hunk 1 above)
Comment (jean-philippe-martin) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: Certainly! I named it NIO_MAX_REOPENS since it also controls how many times we re-open the channel (and normal retries are handled differently).

**Comment by jean-philippe-martin** [comment]
`getFeatureReader` uses `IOUtils.getPath` so it's fine, this sets the retries.

**Comment by jean-philippe-martin** [comment]
The Travis error is:

    The log length has exceeded the limit of 4 MB (this usually means that the test suite is
    raising the same exception over and over).

    The job has been terminated."

Looks like we'll have to make our tests less chatty.

**Comment by lbergelson** [comment]
We're having issues with all of our tests today.  Github is refusing our git-lfs requests because they're over quota and we need to figure out how to either authenticate our requests in a safe way from travis or figure out why we're suddenly going over quota.  It happened very suddenly and I suspect there might be an issue on github's end..

**Comment by droazen** [comment]
Unfortunately the only thing to do for now is to keep hitting "restart" on all the travis builds until they pass..

**Comment by jean-philippe-martin** [comment]
One problem I've noticed is that the log is full of notices from tests we're skipping. The limit on Travis log length is new I think. Removing the mention of each individual skipped may help us have logs that fall below Travis' maximum.

**Comment by droazen** [comment]
The only reason the logs are so long right now is because git-lfs is failing intermittently to download the large files. Under normal circumstances we definitely want to see in the logs when a test is skipped, as there should not be many (or any) skipped tests.

**Comment by jean-philippe-martin** [comment]
I can put this back, but I do think that many tests are skipped by one or the other of our 3 runs. Consider this line from the [log](https://s3.amazonaws.com/archive.travis-ci.org/jobs/219089434/log.txt):

     194479 tests completed, 7 failed, 9729 skipped

**Comment by droazen** [comment]
That is not normal @jean-philippe-martin -- eg., a recent passing build of the unit tests on master (https://travis-ci.org/broadinstitute/gatk/builds/219053956) has just 5 out of 419980 tests skipped:

```
Results: SUCCESS (419980 tests, 419975 successes, 0 failures, 5 skipped)
```

The ~9000+ skips you're seeing are almost certainly due to the intermittent git lfs quota issues -- we've contacted travis (and github) support about this. If you keep re-running the tests they should pass eventually.

**Comment by jean-philippe-martin** [comment]
I see, I was mistaken. Thank you @droazen! I'll retry instead, then.

**Comment by jean-philippe-martin** [comment]
There we go! Thank you!
